src/states/node.rs

```diff
@@ -856,78 +851,72 @@ impl Node {
     }
 
     fn add_to_routing_table(&mut self, public_id: PublicId, peer_id: PeerId) {
-        let name = *public_id.name();
-        if self.peer_mgr.routing_table().contains(&name) {
-            return; // We already sent a `NodeIdentify` to this peer.
-        }
-
-        let bucket_index = self.name().bucket_index(&name);
-        let common_groups = self.peer_mgr
-            .routing_table()
-            .is_in_any_close_group_with(bucket_index, GROUP_SIZE);
-
         match self.peer_mgr.add_to_routing_table(public_id, peer_id) {
-            None => {
-                debug!("{:?} Peer was not added to the routing table: {:?}",
+            Err(RoutingTableError::AlreadyExists) => return,  // already in RT
+            Err(error) => {
+                debug!("{:?} Peer {:?} was not added to the routing table: {}",
                        self,
-                       peer_id);
+                       peer_id,
+                       error);
                 self.disconnect_peer(&peer_id);
             }
-            Some(AddedNodeDetails { must_notify, unneeded }) => {
-                info!("{:?} Added {:?} to routing table.", self, name);
+            Ok(split_prefix) => {
+                info!("{:?} Added {:?} to routing table.", self, public_id.name());
                 if self.peer_mgr.routing_table().len() == 1 {
                     self.send_event(Event::Connected);
                 }
-                for peer_id in self.peer_mgr.get_peer_ids(&must_notify) {
-                    let message = DirectMessage::NewNode(public_id);
-                    let _ = self.send_direct_message(&peer_id, message);
-                }
-                for peer_id in self.peer_mgr.get_peer_ids(&unneeded) {
-                    let message = DirectMessage::ConnectionUnneeded(*self.name());
-                    let _ = self.send_direct_message(&peer_id, message);
+
+                if let Some(prefix) = split_prefix {
+                    // None of these will have been in our group, so no need to notify Routing user.
+                    let peers_to_drop = self.peer_mgr.split_group(prefix);
+                    for peer_id in peers_to_drop {
+                        let _ = self.crust_service.disconnect(peer_id);
+                    }
                 }
 
-                self.reset_bucket_refresh_timer();
+                let all_rt_contacts = self.peer_mgr
+                    .routing_table()
+                    .iter()
+                    .filter(|name| *name != public_id.name())
+                    .cloned()
+                    .collect();
+                let peer_ids_to_notify = self.peer_mgr.get_peer_ids(&all_rt_contacts);
```

Comment:
I'm slightly in favour of changing it, but I don't insist. :-)

And the borrow checker should also be happy, because neither call is `mut`.